### PR TITLE
Added processing of NO_PROXY variable

### PR DESCRIPTION
**Pull request by EzzhevNikita**
Process NO_PROXY env variable before passing to the agent.proxybypasslist. Since we are passing hosts to the regular expression we need to escape dots as special symbols that are presented in hosts.

By new logic:
* if the host has already escaped they won't be escaped one more time
* all the unescaped dots in host templates will be
* wildcard - '.*' that could exist in host template in NO_PROXY

### CODE REVIEW

File: src/Microsoft.VisualStudio.Services.Agent/VstsAgentWebProxy.cs [hunk 1]
```diff
@@ -141,8 +141,15 @@ public void LoadProxyBypassList()
...
             foreach (string bypass in proxyBypassEnv.Split(new [] {',', ';'}).Where(value => !string.IsNullOrWhiteSpace(value)).Select(value => value.Trim()))
             {
-                Trace.Info($"Bypass proxy for: {bypass}.");
-                ProxyBypassList.Add(bypass);
+                string saveRegexString = bypass;
+
+                var regExp = new Regex("(?<!\\\\)([.])(?!\\*)");
+                var replace = "\\$1";
+
+                saveRegexString = regExp.Replace(saveRegexString, replace);
```
Comment (alexander-smolyakov) [inline src/Microsoft.VisualStudio.Services.Agent/VstsAgentWebProxy.cs:149]: Could it make sense to extract this logic into a separate method and add documentation regarding why we are processing the list of proxies in that way?

File: src/Microsoft.VisualStudio.Services.Agent/VstsAgentWebProxy.cs [hunk 2]
```diff
@@ -141,8 +141,15 @@ public void LoadProxyBypassList()
 
             foreach (string bypass in proxyBypassEnv.Split(new [] {',', ';'}).Where(value => !string.IsNullOrWhiteSpace(value)).Select(value => value.Trim()))
             {
-                Trace.Info($"Bypass proxy for: {bypass}.");
-                ProxyBypassList.Add(bypass);
+                string saveRegexString = bypass;
+
+                var regExp = new Regex("(?<!\\\\)([.])(?!\\*)");
```
Comment (alexander-smolyakov) [inline src/Microsoft.VisualStudio.Services.Agent/VstsAgentWebProxy.cs:146]: I believe we should cover this new processing logic by unit tests.

File: src/Microsoft.VisualStudio.Services.Agent/VstsAgentWebProxy.cs (hunk 1 above)
Comment (EzzhevNikita) [reply src/Microsoft.VisualStudio.Services.Agent/VstsAgentWebProxy.cs]: done

File: src/Microsoft.VisualStudio.Services.Agent/VstsAgentWebProxy.cs (hunk 2 above)
Comment (EzzhevNikita) [reply src/Microsoft.VisualStudio.Services.Agent/VstsAgentWebProxy.cs]: Added test for new logic

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks!

**Review by alexander-smolyakov — APPROVED** [review]
LGTM, thanks!

**Comment by alexander-smolyakov** [comment]
Tested these changes locally - LGTM.

`NO_PROXY` variable value: 
`127.0.0.1,.ing.net,.intranet,.corp.int,.*corp.int,127\\.0\\.0\\.1,*.net.*.,*.*.*.*.,.\asdasd.\asdasda.\asdasd,*\.\.*,123/././.****,.....,././././././,,
`

From `Agent.Worker` logs:
```
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Config proxy use DefaultNetworkCredentials.
[2021-08-11 07:06:45Z INFO HostContext] Well known directory 'Bin': 'D:\@Repos\azure-pipelines-agent\_layout\win-x64\bin'
[2021-08-11 07:06:45Z INFO HostContext] Well known directory 'Root': 'D:\@Repos\azure-pipelines-agent\_layout\win-x64'
[2021-08-11 07:06:45Z INFO HostContext] Well known config file 'ProxyBypass': 'D:\@Repos\azure-pipelines-agent\_layout\win-x64\.proxybypass'
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: 127\.0\.0\.1.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: \.ing\.net.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: \.intranet.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: \.corp\.int.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: .*corp\.int.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: 127\\.0\\.0\\.1.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: *\.net.*\..
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: *.*.*.*\..
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: \.\asdasd\.\asdasda\.\asdasd.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: *\.\.*.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: 123/\./\./.****.
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: \.\.\.\.\..
[2021-08-11 07:06:45Z INFO VstsAgentWebProxy] Bypass proxy for: \./\./\./\./\./\./.
```

**Pull request merged**